docs/layout-of-source-files.rst

```diff
@@ -216,101 +216,562 @@ the code below creates new global symbols ``alias`` and ``symbol2`` which refere
+       If a contract imports a file that is not present in `sources`, the compiler will use the file
+       loader as in any other situation, which may result in the source being read from disk.
+
+#. **Standard JSON (as URL)**
+
+   When using Standard JSON it is possible to  tell the compiler to load the files from disk directly:
+
+   .. code-block:: json
+
+       {
+           "language": "Solidity",
+           "sources": {
+               "/usr/local/dapp-bin/token.sol": {
+                   "urls": ["/projects/mytoken.sol"]
+               }
+           },
+           "settings": {"outputSelection": {"*": { "*": ["metadata", "evm.bytecode"]}}}
+       }
+
+   The path specified in ``urls`` is only passed to the file loader and used to locate the file.
+   It does not affect the import key and is not included in contract metadata.
+
+   Paths in ``urls`` are affected by base path and any other transformations performed by the file loader.
+
+#. **Standard input**
+
+   The last way to provide the source is by sending it to compiler's standard input:
+
+   .. code-block:: bash
+
+       echo 'import "./util.sol"; contract C {}' | solc -
+
+   The content of the standard input is identified in the virtual filesystem by a special import key:
+   ``<stdin>``.
+
+.. warning::
+
+    The compiler uses import keys to determine whether imports refer to the same module or not.
+    If you refer to a file in multiple ways that translate to different keys, it will be compiled
+    multiple times.
+
+    For example:
+
+    .. code-block:: solidity
+        :caption: /code/contracts/contract.sol
+
+        import "tokens/token.sol"   // Import key: tokens/token.sol
+        import "tokens///token.sol" // Import key: tokens///token.sol
+
+    .. code-block:: bash
+
+        cd /code
+        solc tokens/contract.sol /code/tokens/token.sol # Import key: /code/tokens/token.sol
+
+    In the above ``token.sol`` will end up in the virtual filesystem under three different
+    import keys even though all the paths refer to the same file in the underlying filesystem.
+
+    To avoid this situation it is recommended to always use the canonical form of paths in your
+    imports and to only list the top-level files that are not imported by other files when
+    invoking the CLI compiler.
+
+Now that we know how the virtual filesystem works, let us go through the rules used to translate
+import paths into import keys in more detail.
+
+.. index:: absolute import
+
+Absolute Imports
+~~~~~~~~~~~~~~~~
+
+An *absolute import* always starts with a forward slash (``/``).
+The import path translates directly to an import key without normalization of any kind:
+
+::
+
+    import "/project/lib/util.sol" as util;          // Import key: /project/lib/util.sol
+    import "/project/lib/../lib///math.sol" as math; // Import key: /project/lib/../lib///math.sol
+
+In the above you might expect the import key to be reduced to ``/project/lib/math.sol`` but it is
```

Comment:
I think we should fix this,shouldn't we?

Reply:
As long as we agree that it should not work that way. From #2738 I got the impression that things are not being normalized on purpose.